src/persist/src/indexed/columnar.rs

```diff
@@ -405,6 +444,93 @@ impl ColumnarRecordsBuilder {
     }
 }
 
+/// A new-type so we can impl FromIterator for Vec<ColumnarRecords>.
+#[derive(Debug)]
+pub struct ColumnarRecordsVec(pub Vec<ColumnarRecords>);
+
+impl ColumnarRecordsVec {
+    /// Unwraps this ColumnarRecordsVec, returning the underlying
+    /// Vec<ColumnarRecords>.
+    pub fn into_inner(self) -> Vec<ColumnarRecords> {
+        self.0
+    }
+}
+
+/// A wrapper around ColumnarRecordsBuilder that chunks as necessary to keep
+/// each ColumnarRecords within the required size bounds.
+#[derive(Debug)]
+pub struct ColumnarRecordsVecBuilder {
+    current: ColumnarRecordsBuilder,
+    filled: Vec<ColumnarRecords>,
+    // Defaults to the KEY_VAL_DATA_MAX_LEN const but override-able for testing.
+    key_val_data_max_len: usize,
+}
+
+impl Default for ColumnarRecordsVecBuilder {
+    fn default() -> Self {
+        Self {
+            current: ColumnarRecordsBuilder::default(),
+            filled: Vec::with_capacity(1),
+            key_val_data_max_len: KEY_VAL_DATA_MAX_LEN,
+        }
+    }
+}
+
+impl ColumnarRecordsVecBuilder {
+    #[cfg(test)]
+    fn new_with_len(key_val_data_max_len: usize) -> Self {
+        assert!(key_val_data_max_len <= KEY_VAL_DATA_MAX_LEN);
+        ColumnarRecordsVecBuilder {
+            current: ColumnarRecordsBuilder::default(),
+            filled: Vec::with_capacity(1),
+            key_val_data_max_len,
+        }
+    }
+
+    /// The number of (potentially duplicated) ((Key, Val), Time, Diff) records
+    /// stored in Self.
+    pub fn len(&self) -> usize {
+        self.current.len() + self.filled.iter().map(|x| x.len()).sum::<usize>()
+    }
+
+    /// Reserve space for `additional` more records, based on `key_size_guess` and
+    /// `val_size_guess`.
+    ///
+    /// The guesses for key and val sizes are best effort, and if they end up being
+    /// too small, the underlying buffers will be resized.
+    pub fn reserve(&mut self, additional: usize, key_size_guess: usize, val_size_guess: usize) {
+        // TODO: This logic very much breaks down if we do end up having to
```

Comment:
just to double check this todo - the logic breaks down in that we reserve individually min(max len, additional * key_size guess) for the keys and min(max len, additional * val_size guess) for the vals

more specifically it seems like theres two breakages:
1. we won't always reserve for as much as you ask for
2. if you ask for say 1 thousand 20 MB keys, and 1 thousand 1 MB values - we'll allocate a `ColumnarRecords` that has space for 100 keys and 1000 values, where 90% of the allocated space for values is a memory leak?

Reply:
this is all fine for now - i think maybe its worth elaborating in the todo?